File: app/src/main/java/kandroid/data/RawData.java
package kandroid.data;

import java.io.*;
import java.net.*;
import java.util.*;
import java.util.zip.*;
import kandroid.observer.*;
import kandroid.utils.*;
import org.apache.commons.io.*;

public class RawData implements Runnable {

    private final String uri;
    private final byte[] request;
    private final byte[] response;
    private final Map<String, String> postField;

    public RawData(String uri, byte[] request, byte[] response) {
        this.uri = uri;
        this.request = request;
        this.response = response;
        this.postField = null;
    }

    public RawData(String uri, Map<String, String> postField, byte[] response) {
        this.uri = uri;
        this.request = null;
        this.response = response;
        this.postField = postField;
    }

    public String getUri() {
        return uri;
    }

    public byte[] getResponse() {
        return response;
    }

    public RawData decode() {
        if (response.length > 0) {
            try {
                Map<String, String> field = null;
                if (this.request != null) {
                    field = getQueryMap(URLDecoder.decode(new String(this.request).trim(), "UTF-8"));
                }

                InputStream stream = new ByteArrayInputStream(this.response);

                // Check GZIP Magic Number
                if ((this.response[0] == (byte) 0x1f) && (this.response[1] == (byte) 0x8b)) {
                    stream = new GZIPInputStream(stream);
                }
                // Remove "svdata="
                int read;
                while (((read = stream.read()) != -1) && (read != '=')) {
                }

                byte[] decodedData = IOUtils.toByteArray(stream);

                return new RawData(getUri().substring(8), field, decodedData) {
                    @Override
                    public String toString() {
                        return new String(getResponse());
                    }
                };
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return this;
    }

    private static Map<String, String> getQueryMap(String query) {
        String[] params = query.split("&");
        Map<String, String> map = new HashMap<>();
        for (String param : params) {
            String[] splited = param.split("=");
            String name = splited[0];
            String value = null;
            if (splited.length == 2) {
                value = splited[1];
            }
            map.put(name, value);
        }
        return map;
    }
	
	@Override
	public void run() {
		RawData rawData = decode();
		ApiBase api = ApiLoader.get(rawData.getUri());
		Utils.requireNonNull(api).receivedData(rawData);
	}
}

File: core/src/main/java/kandroid/observer/RawData.kt
package kandroid.observer

import kandroid.config.Config
import kandroid.utils.log.Logger
import kandroid.utils.yyyyMMdd_HHmmssSSS
import org.apache.commons.io.FileUtils
import org.apache.commons.io.IOUtils
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.net.URLDecoder
import java.nio.charset.Charset
import java.util.*
import java.util.zip.GZIPInputStream

abstract class RawData(val uri: String,
                       val date: Date = Date(),
                       val isFromKcServer: Boolean) {

    abstract val requestString: String

    abstract val responseString: String

    val requestMap: MutableMap<String, String> by lazy {
        val map = HashMap<String, String>()
        val params = URLDecoder.decode(requestString, "UTF-8").split("&")
        for (param in params) {
            val split = param.split("=")
            if (split.size == 2) {
                map.put(split[0], split[1])
            }
        }
        map
    }

    open fun decode(): RawData = this

    fun saveToFile() {
        if (!isFromKcServer) return

        if (Config.isSaveKcsApi) {
            val dateString: String = date.yyyyMMdd_HHmmssSSS
            try {
                if (Config.isSaveKcsRequest) {
                    val fileName = "${dateString}Q@${uri.replace('/', '@')}.txt"
                    val file = Config.getSaveKcsApiFile(fileName)
                    FileUtils.writeStringToFile(file, requestString, Charset.defaultCharset())
                }
            } catch (e: Exception) {
                Logger.e("保存Request失败 -> $e", e)
            }

            try {
                if (Config.isSaveKcsResponse) {
                    val fileName = "${dateString}S@${uri.replace('/', '@')}.json"
                    val file = Config.getSaveKcsApiFile(fileName)
                    FileUtils.writeStringToFile(file, responseString, Charset.defaultCharset())
                }
            } catch (e: Exception) {
                Logger.e("保存Response失败 -> $e", e)
            }
        }
    }
}

class ByteArrayRawData(uri: String,
                       val request: ByteArray,
                       val response: ByteArray,
                       date: Date = Date(),
                       isFromKcServer: Boolean = true) : RawData(uri, date, isFromKcServer) {

    override fun decode(): RawData {
        if (response.isEmpty()) return this

        var uri = this.uri
        if (uri.startsWith("/kcsapi/")) {
            uri = uri.substring(8)
        }

        var stream: InputStream = ByteArrayInputStream(this.response)

        // Check GZIP Magic Number
        if (this.response[0] == 0x1f.toByte() && this.response[1] == 0x8b.toByte()) {
            stream = GZIPInputStream(stream)
        }

        // Remove "svdata="
        var read: Int
        do {
            read = stream.read()
        } while (read != -1 && read != '='.toInt())

        val decodedData = IOUtils.toString(stream, Charset.defaultCharset())

        return StringRawData(uri, String(request), decodedData, date, isFromKcServer)
    }

    override val requestString: String by lazy { String(request) }

    override val responseString: String by lazy { String(response) }
}

class StringRawData(uri: String,
                    request: String,
                    response: String,
                    date: Date = Date(),
                    isFromKcServer: Boolean = false) : RawData(uri, date, isFromKcServer) {

    override fun decode(): RawData {
        if (responseString.isEmpty()) return this

        var uri = this.uri
        if (uri.startsWith("/kcsapi/")) {
            uri = uri.substring(8)
        }

        var responseString2: String = responseString

        var stream: GZIPInputStream? = null
        // Check GZIP Magic Number
        if (this.responseString[0] == 31.toChar()) {
            stream = GZIPInputStream(ByteArrayInputStream(responseString.toByteArray()))
        }

        // Remove "svdata="
        if (stream != null) {
            var read: Int
            do {
                read = stream.read()
            } while (read != -1 && read != '='.toInt())

            responseString2 = IOUtils.toString(stream, Charset.defaultCharset())
        } else {
            if (responseString.startsWith("svdata="))
                responseString2 = responseString.substring(7)
        }

        return StringRawData(uri, requestString, responseString2, date, isFromKcServer)
    }

    override val requestString: String = request

    override val responseString: String = response
}